src/clib/Cabinet.h

```diff
@@ -205,4 +206,174 @@ class Cabinet
+     */
+    SharedCabinet() {}
+
+    /**
+     * Add a new object. The index of the object is returned.
+     */
+    static int add(std::shared_ptr<M> obj) {
+        dataRef data = getData();
+        data.push_back(obj);
+        int index = data.size() - 1;
+        return index;
+    }
+
+    /**
+     * Return cabinet size.
+     */
+    static int size() {
+        int size = getData().size();
+        return size;
+    }
+
+    /**
+     * Delete all objects without erasing mapping.
+     */
+    static int clear() {
+        dataRef data = getData();
+        for (size_t i = 0; i < data.size(); i++) {
+            del(i);
+        }
+        return 0;
+    }
+
+    /**
+     * Delete all objects and erase mapping.
+     */
+    static int reset() {
+        getData().clear();
+        return 0;
+    }
+
+    /**
+     * Delete the nth object. After the object is deleted, the pointer to it in the list
+     * is replaced by a pointer to the first element in the list.
+     */
+    static void del(size_t n) {
+        dataRef data = getData();
+        if (n >= 0 && n < data.size()) {
+            data[n].reset();
+        } else {
+            throw CanteraError("SharedCabinet::del",
+                "Attempt made to delete a non-existing object.");
+        }
+    }
+
+    /**
+     * Return a shared pointer to object n.
+     */
+    static std::shared_ptr<M>& at(size_t n) {
+        dataRef data = getData();
+        if (n < 0 || n >= data.size()) {
+            throw CanteraError("SharedCabinet::at", "Index {} out of range.", n);
+        }
+        return data[n];
+    }
+
+    /**
+     * Return a reference to object n.
+     */
+    static M& item(size_t n) {
+        auto ptr = at(n);
+        if (!ptr) {
+            throw CanteraError("SharedCabinet::item",
+                "Object with index {} has been deleted.", n);
+        }
+        return *ptr;
+    }
+
+    /**
+     * Return a reference to object n, cast to a reference of the specified type.
+     */
+    template <class T>
+    static T& get(size_t n) {
+        auto obj = std::dynamic_pointer_cast<T>(at(n));
+        if (obj) {
+            return *obj;
+        }
+        throw CanteraError("SharedCabinet::get", "Item is not of the correct type.");
+    }
+
+    /**
+     * Return the index in the SharedCabinet to the specified object, or -1 if the
+     * object is not in the SharedCabinet.
+     */
+    static int index(const M& obj) {
+        dataRef data = getData();
+        int count = 0;
+        for (const auto& item : data) {
+            if (item.get() == &obj) {
+                return count;
+            }
+            count++;
+        }
+        return -1;
+    }
+
+private:
+    /**
+     * Static function that returns a pointer to the data member of
+     * the singleton SharedCabinet<M> instance. All member functions should
+     * access the data through this function.
+     */
+    static dataRef getData() {
+        if (s_storage == 0) {
```

Comment:
I don't always remember to do this, but I think it's preferable to write:
```suggestion
        if (s_storage == nullptr) {
```